
# MNIST Dataset Introduction

Most examples are using MNIST dataset of handwritten digits. The dataset contains 60,000 examples for training and 10,000 examples for testing. The digits have been size-normalized and centered in a fixed-size image (28x28 pixels) with values from 0 to 1. For simplicity, each image has been flatten and converted to a 1-D numpy array of 784 features (28*28).

## Overview

![MNIST Digits](http://neuralnetworksanddeeplearning.com/images/mnist_100_digits.png)

## Usage
In our examples, we are using TensorFlow [input_data.py](https://github.com/tensorflow/tensorflow/blob/r0.7/tensorflow/examples/tutorials/mnist/input_data.py) script to load that dataset.
It is quite useful for managing our data, and handle:

- Dataset downloading

- Loading the entire dataset into numpy array: 




In [0]:
import tensorflow as tf

%tensorflow_version 2.x

Load and prepare the MNIST dataset. Convert the samples from integers to floating-point numbers:

In [8]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

11493376/11490434 [==============================] - 0s 0us/step


Build the `tf.keras.Sequential` model by stacking layers. Choose an optimizer and loss function for training:

In [0]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

Train and evaluate the model:

In [10]:
model.fit(x_train, y_train, epochs=5)

model.evaluate(x_test,  y_test, verbose=2)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 5s 83us/sample - loss: 0.2941 - accuracy: 0.9136
Epoch 2/5
60000/60000 [==============================] - 4s 72us/sample - loss: 0.1458 - accuracy: 0.9566
Epoch 3/5
60000/60000 [==============================] - 4s 71us/sample - loss: 0.1093 - accuracy: 0.9668
Epoch 4/5
60000/60000 [==============================] - 4s 71us/sample - loss: 0.0891 - accuracy: 0.9725
Epoch 5/5
60000/60000 [==============================] - 4s 73us/sample - loss: 0.0739 - accuracy: 0.9770
10000/10000 - 0s - loss: 0.0712 - accuracy: 0.9788


[0.0712438988886308, 0.9788]

The image classifier is now trained to ~98% accuracy on this dataset. To learn more, read the TensorFlow tutorials.